### PR TITLE
Refactor `caseCon` and add more documentation

### PR DESCRIPTION
**Pull request by christiaanb**
While trying to determine whether `caseCon` has any lingering variable-capture issues I found the code somewhat opaque, so I added some comments, and also shared some of the WHNF logic with `reduceConst`

### CODE REVIEW

File: clash-lib/src/Clash/Normalize/Transformations.hs [hunk 1]
```diff
@@ -494,150 +490,150 @@ inlineNonRep _ e = return e
...
+      -- or a primitive for which the evaluator doesn't have any evaluation
+      -- rules.
+      _ -> do
+        tcm <- Lens.view tcCache
+        let subjTy = termType tcm subj
+        tran <- Lens.view typeTranslator
+        reprs <- Lens.view customReprs
+        case (`evalState` HashMapS.empty) (coreTypeToHWType tran reprs tcm subjTy) of
+          Right (FilteredHWType (Void (Just hty)) _areVoids)
+            | hty `elem` [BitVector 0, Unsigned 0, Signed 0, Index 1]
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize/Transformations.hs]: Couldn't we make this a bit more general and use `typeSize`?

File: clash-lib/src/Clash/Normalize/Transformations.hs [hunk 2]
```diff
@@ -494,150 +490,150 @@ inlineNonRep _ e = return e
...
+            [] -> altE
+            _  ->
+              -- See Note [CaseCon deshadow]
+              let ((is3,substIds),binds2) =
+                    List.mapAccumL newBinder (is1,[]) binds1
+                  subst = extendIdSubstList (mkSubst is3) substIds
+                  body  = substTm "caseCon0" subst altE
+              in  case Maybe.catMaybes binds2 of
+                    []     -> body
+                    binds3 -> Letrec binds3 body
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize/Transformations.hs]: No change requested, but I've found a good way of formatting `let` bindings without indenting "too much" is:

```
let
  a = 123
  b = 456
in
  a + b
```

File: clash-lib/src/Clash/Normalize/Transformations.hs [hunk 3]
```diff
@@ -494,150 +490,150 @@ inlineNonRep _ e = return e
...
+      -- Use the original inScopeSet 'is0' here, not the extended inScopeSet
+      -- 'is1', otherwise we'd make the "caseCon1" substitution substitute
+      -- free variables that were shadowed by the pattern!
+      let subst = extendTvSubstList (mkSubst is0)
+                $ zip tvs (drop (length (dcUnivTyVars dc)) (Either.rights args))
+      changed (substTm "caseCon1" subst altE1)
+    _ -> case alts of
+           ((DefaultPat,altE):_) -> changed altE
+           _ -> changed (undefinedTm ty)
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize/Transformations.hs]: Why is this true? Is it an invariant to have a `DefaultPat` first? And if it's followed by anything, shouldn't that be an error / thrown away by GHC?

Comment (martijnbastiaan) [reply clash-lib/src/Clash/Normalize/Transformations.hs]: Whatever the answer is, maybe it should be added as a comment.

File: clash-lib/src/Clash/Normalize/Transformations.hs (hunk 1 above)
Comment (christiaanb) [reply clash-lib/src/Clash/Normalize/Transformations.hs]: No... we really only want to do it for those types; so we'd need to extract the `hty` out of the `Void`.

**Pull request merged**